src/cockpit/config.py

```diff
@@ -15,10 +15,65 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
+import configparser
+import logging
 import os
 
 from pathlib import Path
 
+from systemd_ctypes import bus
+
+logger = logging.getLogger(__name__)
+
 ETC_COCKPIT = Path('/etc/cockpit')
 XDG_CONFIG_HOME = Path(os.getenv('XDG_CONFIG_HOME') or os.path.expanduser('~/.config'))
 DOT_CONFIG_COCKPIT = XDG_CONFIG_HOME / 'cockpit'
+
+
+class Config(bus.Object, interface='cockpit.Config'):
+    def __init__(self):
+        self.reload()
+
+    @bus.Interface.Method(out_types='s', in_types='ss')
+    def get_string(self, section, key):
+        try:
+            return self.config[section][key]
+        except KeyError:
+            raise bus.BusError('cockpit.Config.KeyError', f'key {key} in section {section} does not exist')
+
+    @bus.Interface.Method(out_types='u', in_types='ssuuu')
+    def get_u_int(self, section, key, default, maximum, minimum):
+        try:
+            value = self.config[section][key]
+        except KeyError:
+            return default
+
+        try:
+            int_val = int(value)
+        except ValueError:
+            logger.warning(f'cockpit.conf: [{section}] {key} is not an integer')
+            return default
+
+        return min(max(int_val, minimum), maximum)
+
+    @bus.Interface.Method()
+    def reload(self):
+        self.config = configparser.ConfigParser(interpolation=None)
+
+        config_dirs = os.environ.get('XDG_CONFIG_DIRS', '/etc').split(':')
+        for config_dir in config_dirs:
+            config_file = os.path.join(config_dir, 'cockpit', 'cockpit.conf')
+            if os.path.exists(config_file):
+                logger.debug("cockpit.Config: loading %s", config_file)
+                self.config.read(config_file)
+                break
+
+        # it's ok to not have a config file and thus leave self.config empty
+
+
+class Environment(bus.Object, interface='cockpit.Environment'):
+    variables = bus.Interface.Property('a{ss}')
+
+    @variables.getter
+    def get_variables(self):
+        return os.environ.copy()
```

Comment:
I was thinking about this since yesterday, and how passing `os.environ` itself would be a weird choice, but that making a copy would also be strange.  It's annoying that Python doesn't really have value types, and that everything is a reference to mutable state... but that's Python.

In any case, I think the cautious approach is also completely fine.

Reply:
In any case, this is much much better.  Thanks!  And thanks for the tests.